modules/message-lane/src/instant_payments.rs

```diff
@@ -55,34 +65,252 @@ where
+	}
+}
+
+/// Pay rewards to given relayers, optionally rewarding confirmation relayer.
+fn pay_relayers_rewards<AccountId, Balance, Currency>(
+	currency: &mut Currency,
+	confirmation_relayer: &AccountId,
+	relayers_rewards: RelayersRewards<AccountId, Balance>,
+	relayer_fund_account: &AccountId,
+	confirmation_fee: Balance,
+) where
+	AccountId: Debug + Default + Encode + PartialEq,
+	Balance: Debug + AtLeast32BitUnsigned + From<MessageNonce> + Copy,
+	Currency: InstantCurrency<AccountId, Balance>,
+{
+	// reward every relayer, excepting `confirmation_relayer`
+	let mut confirmation_relayer_reward = Balance::zero();
+	for (relayer, reward) in relayers_rewards {
+		let mut relayer_reward = reward.reward;
+
+		if relayer != *confirmation_relayer {
+			// If delivery confirmation is submitted by other relayer, let's deduct confirmation fee
+			// from relayer reward.
+			//
+			// If confirmation fee has been increased (or if it was the only component of message fee),
+			// then messages relayer may receive zero reward.
+			let mut confirmation_reward = confirmation_fee.saturating_mul(reward.messages.into());
+			if confirmation_reward > relayer_reward {
+				confirmation_reward = relayer_reward;
+			}
+			relayer_reward = relayer_reward.saturating_sub(confirmation_reward);
+			confirmation_relayer_reward = confirmation_relayer_reward.saturating_add(confirmation_reward);
+		} else {
+			// If delivery confirmation is submitted by this relayer, let's add confirmation fee
+			// from other relayers to this relayer reward.
+			confirmation_relayer_reward = confirmation_relayer_reward.saturating_add(reward.reward);
+			continue;
+		}
+
+		pay_relayer_reward(currency, relayer_fund_account, &relayer, relayer_reward);
+	}
+
+	// finally - pay reward to confirmation relayer
+	pay_relayer_reward(
+		currency,
+		relayer_fund_account,
+		confirmation_relayer,
+		confirmation_relayer_reward,
+	);
+}
+
+/// Transfer funds from relayers fund account to given relayer.
+fn pay_relayer_reward<AccountId, Balance, Currency>(
+	currency: &mut Currency,
+	relayer_fund_account: &AccountId,
+	relayer_account: &AccountId,
+	reward: Balance,
+) where
+	AccountId: Debug,
+	Balance: Debug + Zero + Copy,
+	Currency: InstantCurrency<AccountId, Balance>,
+{
+	if reward.is_zero() {
+		return;
+	}
+
+	let pay_result = currency.transfer(relayer_fund_account, relayer_account, reward);
 
-		// we can't actually do anything here, because rewards are paid as a part of unrelated transaction
-		match pay_result {
-			Ok(_) => frame_support::debug::trace!(
-				target: "runtime",
-				"Rewarded relayer {:?} with {:?}",
-				relayer,
-				reward,
-			),
-			Err(error) => frame_support::debug::trace!(
-				target: "runtime",
-				"Failed to pay relayer {:?} reward {:?}: {:?}",
-				relayer,
-				reward,
-				error,
-			),
+	// we can't actually do anything here, because rewards are paid as a part of unrelated transaction
```

Comment:
What do you mean by "we can't actually do anything here"?

Reply:
We can't handle this error e.g. by failing whole transaction. So the only thing that is left is to log it.

Reply:
Can you add that clarification as part of the comment?